lib/Service/ScheduledMessageService.php

```diff
@@ -0,0 +1,258 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Talk\Service;
+
+use OCA\Talk\Chat\CommentsManager;
+use OCA\Talk\Chat\MessageParser;
+use OCA\Talk\Exceptions\ScheduledMessagingNotAllowedException;
+use OCA\Talk\Model\Message;
+use OCA\Talk\Model\ScheduledMessage;
+use OCA\Talk\Model\ScheduledMessageMapper;
+use OCA\Talk\Model\Thread;
+use OCA\Talk\Participant;
+use OCA\Talk\Room;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\Comments\IComment;
+use OCP\DB\Exception;
+use OCP\IL10N;
+use Psr\Log\LoggerInterface;
+
+class ScheduledMessageService {
+	public const MAX_CHAT_LENGTH = 32000;
+
+	public function __construct(
+		private readonly ScheduledMessageMapper $scheduledMessageMapper,
+		protected ThreadService $threadService,
+		protected ITimeFactory $timeFactory,
+		protected IL10N $l,
+		protected LoggerInterface $logger,
+		protected CommentsManager $commentsManager,
+		protected MessageParser $messageParser,
+	) {
+	}
+
+	public function scheduleMessage(
+		Room $chat,
+		Participant $participant,
+		string $message,
+		string $messageType,
+		?IComment $parent,
+		int $threadId,
+		\DateTime $sendAt,
+		array $metadata = [],
+	): ScheduledMessage {
+		if ($chat->isFederatedConversation()) {
+			$e = new ScheduledMessagingNotAllowedException();
+			$this->logger->error('Attempt to post scheduled message into proxy conversation', ['exception' => $e]);
+			throw $e;
+		}
```

Comment:
Controller takes care of this, easiest for now by not having `#[FederationSupported]` attribute. Then the middleware will take care